### ISSUE TITLE
[HOLD for payment 2023-06-30] [$2000] Web - Split -  The unselected members are selected if you click on Amount field and return back (Works well if you click on description and return back)

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#17662]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to web chrome
2. Go to group chat with 3+ members
3. Click on plus icon and click on split bill
4. Input a number
5. Click on save
6. Remove any member from the list
7. Now click on amount and **Save**
8. Notice that the member you removed earlier is again selected
9. But now , again remove any member and click on description and **Save**
10. Notice that now the problem doesn't occur. The member you removed earlier is still not selected. Thus, the problem only occurs if you click on amount field.

## Expected Result:
The removed member should not be selected on clicking amount field in a similar way how the removed member is not selected when clicked on the description

## Actual Result:
The removed member is selected again if you click on amount field unlike the description field

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.1.3

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/233128095-cedee447-d483-4cfa-b683-d1d295a4279c.mp4

https://user-images.githubusercontent.com/93399543/233127965-d018d429-e9e6-444e-bd1c-e8b11ad80769.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** @priya-zha

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681827527469619

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014c6f23446b0fe33b</li>
        <li>Upwork Job ID: 1648965284444319744</li>
        <li>Last Price Increase: 2023-07-03</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @conorpendergrast (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by conorpendergrast** [comment]
Reproduced, thanks!

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~014c6f23446b0fe33b

**Comment by MelvinBot** [comment]
Current assignee @conorpendergrast is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @thesahindia (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @PauloGasparSv (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by dukenv0307** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Web - Split - The unselected members are selected if you click on Amount field and return back (Works well if you click on description and return back)
### What is the root cause of that problem?
When the user clicks the amount and comes back, `MoneyRequestConfirmationList` Component will be unmounted when clicking the Amount field and then it will be mounted when the user comes back. So that all members will be selected by this logic
https://github.com/Expensify/App/blob/1cd304c4d9871caa13362d76f1a8a11a940306b7/src/components/MoneyRequestConfirmationList.js#L88-L90
This problem does not happen when clicking the description and return back because when clicking the description, `MoneyRequestConfirmationList` Component will not be unmounted
### What changes do you think we should make in order to solve the problem?
We should save the selected field of participants in ONYX

### What alternative solutions did you explore? (Optional)

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
All participants selected state is lost when we navigate to edit amount page.

### What is the root cause of that problem?
We are rendering conditionally each screen (amount, participants, and confirmation), while the description screen is a separate stack screen. When we navigate from confirmation page to amount page, confirmation page will be unmount, thus the state (selected participants) is lost and everyone becomes selected again.
https://github.com/Expensify/App/blob/1cd304c4d9871caa13362d76f1a8a11a940306b7/src/pages/iou/MoneyRequestModal.js#L414-L447

### What changes do you think we should make in order to solve the problem?

My solution is basically very similar with @huzaifa-99, but because I have spent my time finding the solution, I will still share it.

Currently, we store the selected participants as a state inside `MoneyRequestConfirmationList`. First step, we can introduce a new props called `onParticipantsChange` that will call `setSelectedOptions` to update the selected options/participants in `MoneyRequestModal`, so we don't need the state here anymore. We can also ignore the `getParticipantsWithAmount` because it's redundant. We already call it inside `getSections`
https://github.com/Expensify/App/blob/1cd304c4d9871caa13362d76f1a8a11a940306b7/src/components/MoneyRequestConfirmationList.js#L88-L94
https://github.com/Expensify/App/blob/1cd304c4d9871caa13362d76f1a8a11a940306b7/src/components/MoneyRequestConfirmationList.js#L170-L171
_We can also remove `getParticipantsWithoutAmount` because it won't do nothing anymore since `participants` won't have `descriptiveText` by default._

All `this.state.participants` will be replaced with `this.props.participants`.

To notify `MoneyRequestModal` the new participants, we won't need this `setState` anymore and we will call `onParticipantsChange` here.
https://github.com/Expensify/App/blob/1cd304c4d9871caa13362d76f1a8a11a940306b7/src/components/MoneyRequestConfirmationList.js#L232-L240

Two last things left.

We can look here that we map the list and give it a `selected` value. We will need to do the same when we create the `selectedOptions` array here.
https://github.com/Expensify/App/blob/1cd304c4d9871caa13362d76f1a8a11a940306b7/src/pages/iou/MoneyRequestModal.js#L118-L122

Last, we should reset all `selected` value back to true if the user navigate back from confirm page to amount page.
amount -> confirm -> amount
back, don't reset
amount -> confirm
back, reset

But it's currently not possible because we have this bug https://github.com/Expensify/App/issues/17637. So, we should fix that first before continue this.

### What alternative solutions did you explore? (Optional)
Honestly, I think it would be much cleaner to have this 3 pages (amount, participants, confirmation) have their own stack screen. I don't know the reason why we put all these 3 "pages" inside `MoneyRequestModal` and manually handle the navigation, transition animation, etc. This will become harder to maintain as the component grows. IMO, it would be great if each of those pages have their own Stack.Screen, then we wrap it with a context provider to have a shared value such as the amount, participants, etc. If there are no special reasons to why we put those 3 "pages" in a single screen, I would be happy to do the refactor 😄

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by PauloGasparSv** [comment]
> I prefer @bernhardoj's https://github.com/Expensify/App/issues/17662#issuecomment-1516095431 more because of the code refactor.

I agree with you @thesahindia! So I'm assigning them here : )

> But it's currently not possible because we have this bug https://github.com/Expensify/App/issues/17637. So, we should fix that first before continue this.

Hey @bernhardoj that's a great catch. Can this be worked in parallel while that other bug is fixed? I'm not 100% sure if we should put this on HOLD until that's fixed.

**Comment by bernhardoj** [comment]
As we agree to do the refactor, both of these issues https://github.com/Expensify/App/issues/17637 & https://github.com/Expensify/App/issues/17710 won't happen anymore after the refactor.

Hopefully we are on the same page on the refactoring 😄. I will open the PR soon!

**Comment by bernhardoj** [comment]
Ah, btw I agree with that https://github.com/Expensify/App/issues/17662#issuecomment-1518071661 as we also save description.

**Comment by bernhardoj** [comment]
I keep getting this lint error

```
Oops! Something went wrong! :(

ESLint: 7.32.0

TypeError: externalProps is not iterable (cannot read property undefined)
Occurred while linting /Users/bernhardoj/Documents/dev/freelance/Expensify/src/pages/iou/steps/MoneyRequestAmountPage.js:440
    at /Users/bernhardoj/Documents/dev/freelance/Expensify/node_modules/eslint-config-expensify/eslint-plugin-expensify/onyx-props-must-have-default.js:102:39
```

As it's EOD for me, I will continue to find the fix tomorrow!

**Comment by s77rt** [comment]
What's the state here, are we planing to refactor the request money modal? If so should we hold https://github.com/Expensify/App/issues/17710?

**Comment by dukenv0307** [comment]
I think it would be better to get broader confirmation on whether to do the refactor, there might be conscious reason why it was designed this way to begin with.

**Comment by bernhardoj** [comment]
@s77rt yes, I think we can hold that, unless we cancel the refactor

@PauloGasparSv what do you think about @dukenv0307 comment here https://github.com/Expensify/App/issues/17662#issuecomment-1521165853

While waiting for the answer, I have open the PR for anyone who would like to see the refactor.

**Comment by bernhardoj** [comment]
@PauloGasparSv Bump on this https://github.com/Expensify/App/issues/17662#issuecomment-1521704246 as you previously already agreed with the refactor here https://github.com/Expensify/App/issues/17662#issuecomment-1520125938.

If it's not needed, I think @thesahindia can start reviewing the PR 😄.

**Comment by s77rt** [comment]
I will take it!

**Comment by MelvinBot** [comment]
📣 @s77rt You have been assigned to this job by @PauloGasparSv!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by PauloGasparSv** [comment]
Hey everyone, sorry for the delay here!

> I think it would be better to get broader confirmation on whether to do the refactor, there might be conscious reason why it was designed this way to begin with.

I'm also not comfortable with merging the P.R. without more reviews so I'll start a thread to discuss if we should implement the changes here and to get more reviewers in the P.R.

> As we agree to do the refactor, both of these issues https://github.com/Expensify/App/issues/17637 & https://github.com/Expensify/App/issues/17710 won't happen anymore after the refactor.

Nice! I think it makes sense to put https://github.com/Expensify/App/issues/17710 on Hold for now since https://github.com/Expensify/App/issues/17637 is already on Hold @s77rt

> Hmm, I am not sure about this. We don't clear the description text on the confirmation screen when going back to the amount page so I think we also shouldn't reset the selected value. Also I think it's pretty normal to use the back button when you wanna go back to edit something and it will not be a good experience to see that the changes you made are gone.

I was curious on how it would be implemented and if it would lead to the "Expected Result" so I was waiting to discuss about this in the P.R.
I see @bernhardoj already [agreed](https://github.com/Expensify/App/issues/17662#issuecomment-1520394517) with the comment so I think we are not doing that anymore.

**Comment by PauloGasparSv** [comment]
> @PauloGasparSv Please link to Slack thread about whether we should refactor the page or work on top of what we have.

Sure thing @s77rt I'll start the thread once I'm finish this [deploy blocker](https://github.com/Expensify/App/issues/18145)

**Comment by conorpendergrast** [comment]
@PauloGasparSv Were you able to start that thread?

**Comment by cead22** [comment]
> Hmm, I am not sure about this. We don't clear the description text on the confirmation screen when going back to the amount page so I think we also shouldn't reset the selected value. Also I think it's pretty normal to use the back button when you wanna go back to edit something and it will not be a good experience to see that the changes you made are gone.

I agree with this and not resetting the selections

**Comment by bernhardoj** [comment]
I see we have an announcement on expensify-open-source that we're doing a refactor to money request here https://expensify.slack.com/archives/C01GTK53T8Q/p1683642447007639. Should we hold this one then?

cc: @s77rt @PauloGasparSv

**Comment by s77rt** [comment]
I'm not really sure whether the refactor would cover this. @PauloGasparSv any idea?

**Comment by conorpendergrast** [comment]
@PauloGasparSv Bump on this confirmation please - can you confirm?

**Comment by conorpendergrast** [comment]
DM'ed Paulo now

**Comment by PauloGasparSv** [comment]
@bernhardoj @s77rt Sorry for the delay here!

> I see we have an announcement on expensify-open-source that we're doing a refactor to money request here https://expensify.slack.com/archives/C01GTK53T8Q/p1683642447007639. Should we hold this one then?

**IMO we should have holded for that**. Since most changes related to that announcement are completed in newDot we can move forward with this.

I also don't think we should create a thread to discuss the refactor anymore. I realized that most of the logic for [MoneyRequestModal](https://github.com/Expensify/App/blob/main/src/pages/iou/MoneyRequestModal.js) is pretty recent and was done by @thienlnam (including the routing logic) so I asked them to review the P.R. too! 

@bernhardoj can you please take a look at the P.R. conflicts? I'll start reviewing once that's done : )

**Comment by bernhardoj** [comment]
Got it. Will let you know when it's ready for review.

**Comment by thienlnam** [comment]
Did we ever start that slack thread for seeing if we should refactor this right now? This is a pretty big change and we're also implementing a lot of functionality on top of it right now.

I agree this would be a good change, but not sure if we should do this _right_ now but pinging for thoughts

cc @vitHoracek @luacmartins

**Comment by luacmartins** [comment]
@thienlnam is [this the proposal](https://github.com/Expensify/App/issues/17662#issuecomment-1516095431) we are talking about? If so, it seems pretty self contained to MoneyRequestModal and I think most of the work there is done, so we should be good to work on it.

**Comment by thienlnam** [comment]
We're refactoring the MoneyRequestModal to move the "steps" to an individual screens - but if that's not a worry then we can move forward

**Comment by luacmartins** [comment]
Yea, I think that should be fine as I think we are not doing any more work in that area

**Comment by conorpendergrast** [comment]
PR is still actively under review 👍

**Comment by s77rt** [comment]
Currently we are working on a [bug](https://github.com/Expensify/App/pull/17964#discussion_r1202956927) and we have an ongoing [slack discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1684873343572679) to work on that.

**Comment by conorpendergrast** [comment]
Looks like the PR is still under review! Any issues or trouble there?

**Comment by bernhardoj** [comment]
The discussion just completed and we are continuing the PR now 👍.

**Comment by conorpendergrast** [comment]
Fantastic

**Comment by PauloGasparSv** [comment]
The P.R. is still in progress and has been consistently receiving updates and discussions.

@bernhardoj I don't think we have any blockers here but please LMK if we do

**Comment by bernhardoj** [comment]
Yes, no blockers yet.

**Comment by conorpendergrast** [comment]
[PR](https://github.com/Expensify/App/pull/17964) is still under review!

**Comment by conorpendergrast** [comment]
PR review is ongoing and I see conflicts have recently been resolved

**Comment by conorpendergrast** [comment]
More commits and suggestions in the PR, so it's ongoing!

**Comment by conorpendergrast** [comment]
Ah poop, I didn't hire anyone via the Upwork job (and it has auto-closed). 

Offers sent now:

1. @priya-zha: https://www.upwork.com/nx/wm/offer/25166731
2. @bernhardoj: https://www.upwork.com/nx/wm/offer/25166764
3. @s77rt: https://www.upwork.com/nx/wm/offer/25166777

**Comment by priya-zha** [comment]
Accepted offer. Thanks.

**Comment by s77rt** [comment]
Accepted!

**Comment by bernhardoj** [comment]
Accepted!

Btw, can we ask for a higher bounty as the PR is quite big and solves other issues too? 😄 

@s77rt  Do you agree?

**Comment by conorpendergrast** [comment]
I will escalate that request to discuss in Slack!

**Comment by conorpendergrast** [comment]
Discussing with C+ and Expensifiers in [the C+ channel here](https://expensify.slack.com/archives/C02NK2DQWUX/p1687511487985959)

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: This seems to be an issue that existed since day one but back then there was no option to edit the Amount in the confirmation page but the issue was still reproducible on going back.
- The offending PR has been commented on: n/a
- A discussion in #expensify-bugs has been started: I don't think this is needed here. The regression test below should be enough to prevent this bug from resurfacing.
- Determine if we should create a regression test for this bug: Yes

___

**Regression Test Proposal**
1. Open any group chat
2. Press the grey "+" button in the composer
3. Press Split Bill
4. Enter any amount then press Next
5. Unselect any participant(s)
6. Press Amount and change the amount
7. Verify the unselected participants are still unselected

**Comment by conorpendergrast** [comment]
Regression test request: https://github.com/Expensify/Expensify/issues/296776

**Comment by conorpendergrast** [comment]
All paid, contracts ended, all done. Closing out, thanks everyone
